Reshaping Arrays & Axis Operations

Part 1: Creating a 1D Array

In [1]:
import numpy as np

numbers = np.array([1,2,3,4,5,6,7,8,9,10,11,12])

print(numbers)

[ 1  2  3  4  5  6  7  8  9 10 11 12]


Part 2: reshape()

In [2]:
matrix = numbers.reshape(3,4)
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


Meaning:

3 rows
4 columns

Checking another shape

In [3]:
matrix = numbers.reshape(2,6)
print(matrix)

[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]]


In [4]:
matrix = numbers.reshape(4,3)
print(matrix)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


Rule

The total number of elements must stay the same.

12 elements can become

3 × 4 ✅
2 × 6 ✅
4 × 3 ✅
6 × 2 ✅

Part 3: Automatic Size (-1)

NumPy can calculate one dimension automatically.

In [6]:
matrix = numbers.reshape(3, -1)
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


In [7]:
# or you can do like this.

matrix = numbers.reshape(-1, 2)
print(matrix)

[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]
 [11 12]]


Part 4: Flattening Arrays

Suppose we already have

In [10]:
matrix = np.array([
    [10, 20, 30],
    [40, 50, 60]
])

print(matrix)

# convert back into one row

flat = matrix.flatten()

print(flat)

[[10 20 30]
 [40 50 60]]
[10 20 30 40 50 60]


Part 5: Axis Operations

This is one of the most important NumPy concepts.

marks = np.array([
    [80,90,75],
    [88,92,79],
    [95,85,91]
])

print(marks)

In [ ]:
marks = np.array([
    [80,90,75],
    [88,92,79],
    [95,85,91]
])

print(marks)
print(np.sum(marks, axis=0)) # axis =0 count total downward

[[80 90 75]
 [88 92 79]
 [95 85 91]]
[263 267 245]


80 + 88 + 95 = 263

90 + 92 + 85 = 267

75 + 79 + 91 = 245

axis=1
print(np.sum(marks, axis=1))

Output

[245 259 271]

Explanation

80+90+75

88+92+79

95+85+91

In [16]:
marks = np.array([
    [80,90,75],
    [88,92,79],
    [95,85,91]
])

print(marks)

print(np.sum(marks, axis = 0))

print(np.sum(marks, axis = 1))

print((np.mean(marks, axis= 0)))

print(np.mean(marks, axis = 1))

[[80 90 75]
 [88 92 79]
 [95 85 91]]
[263 267 245]
[245 259 271]
[87.66666667 89.         81.66666667]
[81.66666667 86.33333333 90.33333333]


In [2]:
import numpy as np

sales = np.array([
    [120, 150, 180],
    [200,210,190],
    [170,160,220],
    [190,180,200]
])

print("Sales")
print(sales)

print("\nTotal Sales =", np.sum(sales))

print("\nMonthly Sales")
print(np.sum(sales, axis= 0))

print("\nEmployee Sales")
print(np.sum(sales, axis=1))

print("\nAverage Monthly Sales")
print(np.mean(sales, axis=0))

print("\nHighest Sale Each Month")
print(np.max(sales, axis=0))

print("\nLowest Sale Each Employee")
print(np.min(sales, axis=1))

Sales
[[120 150 180]
 [200 210 190]
 [170 160 220]
 [190 180 200]]

Total Sales = 2170

Monthly Sales
[680 700 790]

Employee Sales
[450 600 550 570]

Average Monthly Sales
[170.  175.  197.5]

Highest Sale Each Month
[200 210 220]

Lowest Sale Each Employee
[120 190 160 180]


PRACTICE EXERCISES


Exercise 1

Create this array:

numbers = np.arange(1,17)

Reshape it into:

4×4
2×8
8×2

In [ ]:
import numpy as np

numbers = np.arange(1,17)

print(numbers.reshape(4,4))

print(numbers.reshape(2,8))

print(numbers.reshape(8,2))


[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]
[[ 1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16]]
[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]
 [11 12]
 [13 14]
 [15 16]]


Exercise 2

Given

scores = np.array([
    [70,80,90],
    [88,77,66],
    [91,83,79]
])

Find:

Total score
Sum of each row
Sum of each column
Average of each row
Average of each column
Highest value in each row
Lowest value in each column

In [18]:
import numpy as np

scores = np.array(
    [[70,80,90],
    [88,77,66],
    [91,83,79]
    ])

print("Total Scores =", np.sum(scores))

print("\nSum of each row")
print(np.sum(scores,axis= 1))

print("\nSum of each column")
print(np.sum(scores, axis= 0))

print("\nAverage of each row")
print(np.mean(scores, axis= 1))

print("\nAverage of each column")
print(np.mean(scores, axis= 0))

print("\nHighest value in each row")
print(np.max(scores, axis= 1))

print("\nLowest value in each column")
print(np.min(scores, axis= 0))

Total Scores = 724

Sum of each row
[240 231 253]

Sum of each column
[249 240 235]

Average of each row
[80.         77.         84.33333333]

Average of each column
[83.         80.         78.33333333]

Highest value in each row
[90 88 91]

Lowest value in each column
[70 77 66]
